src/fdopendir.c

```diff
@@ -22,72 +23,62 @@
 
 #include "common-priv.h"
 
+#include <stdlib.h>
 #include <dirent.h>
 #include <sys/errno.h>
 #include <sys/stat.h>
 #include <unistd.h>
 #include <fcntl.h>
 
+#undef DIR
+#undef opendir
+#undef closedir
+
 
 /*
  * Implementation behavior largely follows these man page descriptions:
  *
  * https://www.freebsd.org/cgi/man.cgi?query=fdopendir&sektion=3
  * https://linux.die.net/man/3/fdopendir
+ *
+ * On success, this function returns allocated memory that must be
+ * deallocated by __mpls_closedir() (see closedir macro in dirent.h).
  */
 
-DIR *fdopendir(int dirfd) {
-    DIR *dir;
+__MP_LEGACY_SUPPORT_DIR *fdopendir(int dirfd) {
+    DIR *dirvp; /* Real DIR ptr */
+    __MP_LEGACY_SUPPORT_DIR *dir; /* Real DIR ptr plus supplied directory file descriptor */
     struct stat st;
-    int oldCWD = -1;
 
-    if (fstat(dirfd, &st) < 0)
-        return 0;
+    /* Check the supplied directory file descriptor */
 
-    if (!S_ISDIR(st.st_mode)) {
-        errno = ENOTDIR;
-        return 0;
-    }
+    if (dirfd != AT_FDCWD) {
+        if (fstat(dirfd, &st) < 0)
+            return 0;
 
-    if (dirfd == AT_FDCWD) {
-        dir = opendir (".");
-        /* dirfd can be closed only upon success */
-        if (dir) PROTECT_ERRNO(close(dirfd));
-        return dir;
+        if (!S_ISDIR(st.st_mode)) {
```

Comment:
Thanks! Another brilliant suggestion. I added more tests to `test/test_fdopendir.c` to make sure that the failure modes (e.g., dirfd=0 and dirfd=-1) behave the same with or without that redundant check, and they do. So now the code looks very lean.

Even better, now that `src/fdopendir.c` doesn't use `struct stat` it no longer has to be compiled multiple times for different architectures. It was the only source file that needed that, so the compilation is simpler. I haven't removed from the `Makefile` the framework for compiling multiple times for different architectures (in case a need for it ever arises again), but it currently isn't used.

That means I was able to move the functions in `src/fdopendir_support.c` into `src/fdopendir.c` (so one file fewer).

I also removed two unused variables in `test/test_fmemopen.c` to get rid of warnings.

And I also added `-Wno-deprecated-declarations` to the `Makefile` to suppress warnings about `lstat64` and `stat64` in a few files that included `<sys/stat.h>` (i.e., `src/atcalls.c` and `test/test_fstatat64.c`). If that's not desirable, I can take it out again.

Reply:
> And I also added `-Wno-deprecated-declarations` to the `Makefile` to suppress warnings about `lstat64` and `stat64` in a few files that included `<sys/stat.h>` (i.e., `src/atcalls.c` and `test/test_fstatat64.c`). If that's not desirable, I can take it out again.

I just have to make sure the compilers we care about (gcc-4.2 and newer, and the system clangs on 10.7+) will swallow that flag.

Reply:
Never mind. It's probably not worth the effort to check. It's just warnings. I've removed it.